### ISSUE TITLE
`stat.Quantile` returns wrong value when dataset is even size

### ISSUE THREAD

**Issue by Iriko0606** [gonum/gonum#1836]
### What are you trying to do?

I'm trying to get median value when I request even size of dataset to `stat.Quantile`

### What did you do?

```
func main() {
	xs := []float64{2, 3, 5, 9}
	median := stat.Quantile(0.5, stat.Empirical, xs, nil)

	fmt.Println(median)
}
```

https://go.dev/play/p/8b6vDMunna8

### What did you expect to happen?

I expect return value `4`

### What actually happened?

its returns `3`

### What version of Go and Gonum are you using?

1.19

### Does this issue reproduce with the current master?
Yes

**Comment by romanwerpachowski** [comment]
The current behaviour is consistent with the documentation. I think you will get the behaviour you want if you call `Quantile` with `stat.LinInterp` instead of `stat.Empirical`.

**Comment by Iriko0606** [comment]
Thanks for your reply.
I've already tried it but I could not get expect value `4`.
Its also returns value `3`

https://go.dev/play/p/lM8ZNR0rTsi

```
func main() {

	xs := []float64{2, 3, 5, 9}
	median := stat.Quantile(0.5, stat.LinInterp, xs, nil)

	fmt.Println(median)
}
```

**Comment by romanwerpachowski** [comment]
I see. I agree the documentation may be understood as promising a different behaviour, but I am not sure if the problem is with the code, or the documentation. @sbinet, what is your opinion?

**Comment by kortschak** [comment]
This is a bug. We say that we are using the same nomenclature as R for cumulant kinds here and this is what R says.

```
> quantile(c(2, 3, 5, 9), probs=0.5, type=7)
50% 
  4 
```

**Comment by Iriko0606** [comment]
Thanks for your investigation and reply.

Is the method should be implemented as Quantile method of another case `Empirical` or `LinInterp`, 
or be implemented as a `Median` method as a new method?

**Comment by kortschak** [comment]
Yes, there should be another CumulantKind. The name than numpy uses is `Internal`, so that's probably a reasonable starting point. I took a look at implementing this the other day, but it's complicated by the impact of weighting.

**Comment by kristofferrandersen** [comment]
Hi @kortschak 
I stumbled upon this while debugging a similar issue to the one mentioned above. Currently, the LinInterp in this package works in a different way than the numpy percentile function using the default method which is linear. 

I would like to contribute with a new linear interpolation method which matches the numpy implementation and also has the common P50 equal to Median feature. 

To me it looks like minimal changes are needed for the unweighted linear interpolation to work.

**Comment by kortschak** [comment]
@kristofferrandersen Are you proposing to change the behaviour of the cumulant kind that we have, or add another? If it's a change in behaviour, I don't think we will accept that. If it's another behaviour, which? This [one](https://github.com/gonum/gonum/issues/1836#issuecomment-1320771896)?

**Comment by kristofferrandersen** [comment]
@kortschak No I agree that we cannot change any of the existing cumulant kinds. 

Referring to [numpys percentile](https://numpy.org/doc/stable/reference/generated/numpy.percentile.html) I think the current "Empirical" implementation covers method 1 "inverted_CDF" for both weighted and unweighted data. I am not completely sure which one of the implementations that the "LinInterp" matches.

I propose we create a new cumulant kind call "linear" (like numpy) which matches the numpy implementation and does not support weights (like the numpy implementation). This implementation will have the advantage that the P50 is equal to the median. We could call it something else which is clearly different from the current "LinInterp".

**Comment by kortschak** [comment]
I would be significantly more inclined towards this if it were implemented to be aware of weights. Without that it's a source of confusion.

**Comment by kristofferrandersen** [comment]
I agree that this would be nice. The reason I propose not to support it is that numpy does not support it.

**Comment by LuizCFdosSantos** [comment]
So don't we need an implementions of something equivalent to the R code:

```
sales_numbers <- c(2, 3, 5, 9)
median(sales_numbers)
```
something like:

```
func main() {
	xs := []float64{2, 3, 5, 9}
	median := stat.Median(xs) //median equals to 4
}
```

**Comment by kortschak** [comment]
@LuizCFdosSantos Yes, but with weights. That is the issue.

**Comment by LuizCFdosSantos** [comment]
@kortschak , but median doesn't have weight: https://en.wikipedia.org/wiki/Median , right? Weight is related to mean instead, I think.

**Comment by kortschak** [comment]
There is no reason not to have a weighted median.
